File: services/order/src/controllers/OrderController.ts
import { Request, Response } from "express";
import { injectable, inject } from "inversify";
import { IOrderService } from "../services/OrderService";
import { ILogger, TYPES } from "@observability/core";
import { OrderStatus } from "../utils/order.utils";
import { v4 as uuidv4 } from "uuid";
import { EventPublisher } from "@observability/common";

@injectable()
export class OrderController {
    constructor(
        @inject("OrderService") private orderService: IOrderService,
        @inject(TYPES.Logger) private logger: ILogger,
        @inject("EventPublisher") private eventPublisher: EventPublisher
    ) {}

    private getTraceId(req: Request): string {
        return (
            (req.headers["x-trace-id"] as string) ||
            (req.headers["X-Trace-Id"] as string) ||
            uuidv4()
        );
    }

    async createOrder(req: Request, res: Response): Promise<void> {
        const traceId = this.getTraceId(req);
        try {
            this.logger.info("Create order request", { traceId, body: req.body });
            const order = await this.orderService.createOrder(req.body, traceId);
            res.status(201).json({
                success: true,
                data: order,
                traceId,
                message: "Order created successfully"
            });
        } catch (error: any) {
            await this.emitErrorAndRespond(res, error, "Failed to create order", traceId, req.body);
        }
    }

    async getOrderById(req: Request, res: Response): Promise<void> {
        const traceId = this.getTraceId(req);
        try {
            const { id } = req.params;
            this.logger.info("Get order by id", { traceId, id });
            const order = await this.orderService.getOrderById(id, traceId);
            if (!order) {
                res.status(404).json({
                    success: false,
                    message: "Order not found",
                    traceId
                });
                return;
            }
            res.json({
                success: true,
                data: order,
                traceId
            });
        } catch (error: any) {
            await this.emitErrorAndRespond(res, error, "Failed to fetch order", traceId, { id: req.params.id });
        }
    }

    async getUserOrders(req: Request, res: Response): Promise<void> {
        const traceId = this.getTraceId(req);
        try {
            const { userId } = req.params;
            this.logger.info("Get user orders", { traceId, userId });
            const orders = await this.orderService.getUserOrders(userId, traceId);
            res.json({
                success: true,
                data: orders,
                count: orders.length,
                traceId
            });
        } catch (error: any) {
            await this.emitErrorAndRespond(res, error, "Failed to fetch user orders", traceId, { userId: req.params.userId });
        }
    }

    async getAllOrders(req: Request, res: Response): Promise<void> {
        const traceId = this.getTraceId(req);
        try {
            this.logger.info("Get all orders", { traceId });
            const orders = await this.orderService.getAllOrders(traceId);
            res.json({
                success: true,
                data: orders,
                count: orders.length,
                traceId
            });
        } catch (error: any) {
            await this.emitErrorAndRespond(res, error, "Failed to fetch orders", traceId, {});
        }
    }

    async updateOrderStatus(req: Request, res: Response): Promise<void> {
        const traceId = this.getTraceId(req);
        try {
            const { id } = req.params;
            const { status } = req.body;
            this.logger.info("Update order status", { traceId, id, status });
            if (!Object.values(OrderStatus).includes(status)) {
                res.status(400).json({
                    success: false,
                    message: "Invalid order status",
                    traceId
                });
                return;
            }
            const order = await this.orderService.updateOrderStatus(id, status, traceId);
            if (!order) {
                res.status(404).json({
                    success: false,
                    message: "Order not found",
                    traceId
                });
                return;
            }
            res.json({
                success: true,
                data: order,
                traceId,
                message: "Order status updated"
            });
        } catch (error: any) {
            await this.emitErrorAndRespond(res, error, "Failed to update order status", traceId, { id: req.params.id, status: req.body.status });
        }
    }

    private async emitErrorAndRespond(
        res: Response,
        error: Error,
        message: string,
        traceId: string,
        context?: any
    ): Promise<void> {
        this.logger.error(message, error, { traceId, error: error.message, stack: error.stack, ...context });
        this.eventPublisher.publishEvent(
            "order.error",
            {
                message,
                error: error.message,
                stack: error.stack,
                context
            },
            traceId
        ).catch((publishErr) => {
            this.logger.error("Failed to publish error event", publishErr, { traceId });
        });

        res.status(500).json({
            success: false,
            message,
            error: error.message,
            traceId
        });
    }
}

File: services-java/order/src/main/java/org/observability/controller/OrderController.java
package org.observability.controller;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.observability.dto.ApiResponse;
import org.observability.dto.CreateOrderDto;
import org.observability.entity.Order;
import org.observability.enums.OrderStatus;
import org.observability.service.EventPublisher;
import org.observability.service.OrderService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/orders")
public class OrderController {

    private static final Logger logger = LoggerFactory.getLogger(OrderController.class);

    private final OrderService orderService;
    private final EventPublisher eventPublisher;

    @Autowired
    public OrderController(OrderService orderService, EventPublisher eventPublisher) {
        this.orderService = orderService;
        this.eventPublisher = eventPublisher;
    }

    @PostMapping
    public ResponseEntity<ApiResponse<Order>> createOrder(
            @Valid @RequestBody CreateOrderDto orderDto,
            HttpServletRequest request) {

        String traceId = getTraceId(request);

        try {
            logger.info("Create order request - TraceId: {}, UserId: {}", traceId, orderDto.getUserId());

            Order order = orderService.createOrder(orderDto, traceId);

            ApiResponse<Order> response = ApiResponse.success(
                    order,
                    "Order created successfully",
                    traceId
            );

            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (Exception e) {
            return emitErrorAndRespond(e, "Failed to create order", traceId, Map.of("userId", orderDto.getUserId()));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<Order>> getOrderById(
            @PathVariable UUID id,
            HttpServletRequest request) {

        String traceId = getTraceId(request);

        try {
            logger.info("Get order by ID - TraceId: {}, OrderId: {}", traceId, id);

            Optional<Order> order = orderService.getOrderById(id, traceId);

            if (order.isEmpty()) {
                ApiResponse<Order> response = ApiResponse.error(
                        "Order not found",
                        "No order found with ID: " + id,
                        traceId
                );
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
            }

            ApiResponse<Order> response = ApiResponse.success(
                    order.get(),
                    null,
                    traceId
            );

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return emitErrorAndRespond(e, "Failed to fetch order", traceId, Map.of("orderId", id));
        }
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<ApiResponse<List<Order>>> getUserOrders(
            @PathVariable String userId,
            HttpServletRequest request) {

        String traceId = getTraceId(request);

        try {
            logger.info("Get user orders - TraceId: {}, UserId: {}", traceId, userId);

            List<Order> orders = orderService.getUserOrders(userId, traceId);

            ApiResponse<List<Order>> response = ApiResponse.success(
                    orders,
                    null,
                    traceId
            );
            response.setCount(orders.size());

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return emitErrorAndRespond(e, "Failed to fetch user orders", traceId, Map.of("userId", userId));
        }
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<Order>>> getAllOrders(HttpServletRequest request) {
        String traceId = getTraceId(request);

        try {
            logger.info("Get all orders - TraceId: {}", traceId);

            List<Order> orders = orderService.getAllOrders(traceId);

            ApiResponse<List<Order>> response = ApiResponse.success(
                    orders,
                    null,
                    traceId
            );
            response.setCount(orders.size());

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return emitErrorAndRespond(e, "Failed to fetch orders", traceId, Map.of());
        }
    }

    @PatchMapping("/{id}/status")
    public ResponseEntity<ApiResponse<Order>> updateOrderStatus(
            @PathVariable UUID id,
            @RequestBody Map<String, String> body,
            HttpServletRequest request) {

        String traceId = getTraceId(request);

        try {
            String statusStr = body.get("status");
            OrderStatus status = OrderStatus.valueOf(statusStr);

            logger.info("Update order status - TraceId: {}, OrderId: {}, Status: {}", traceId, id, status);

            Order order = orderService.updateOrderStatus(id, status, traceId);

            ApiResponse<Order> response = ApiResponse.success(
                    order,
                    "Order status updated successfully",
                    traceId
            );

            return ResponseEntity.ok(response);
        } catch (IllegalArgumentException e) {
            return emitErrorAndRespond(e, "Invalid order status", traceId, Map.of("orderId", id, "status", body.get("status")));
        } catch (Exception e) {
            return emitErrorAndRespond(e, "Failed to update order status", traceId, Map.of("orderId", id));
        }
    }

    private String getTraceId(HttpServletRequest request) {
        String traceId = request.getHeader("x-trace-id");
        if (traceId == null || traceId.isEmpty()) {
            traceId = request.getHeader("X-Trace-Id");
        }
        if (traceId == null || traceId.isEmpty()) {
            traceId = UUID.randomUUID().toString();
        }
        return traceId;
    }

    private <T> ResponseEntity<ApiResponse<T>> emitErrorAndRespond(
            Exception e,
            String message,
            String traceId,
            Map<String, Object> context) {

        logger.error("{} - TraceId: {}, Error: {}", message, traceId, e.getMessage(), e);

        // Publish error event to RabbitMQ
        Map<String, Object> errorPayload = new HashMap<>();
        errorPayload.put("message", message);
        errorPayload.put("error", e.getMessage());
        errorPayload.put("stack", getStackTrace(e));
        errorPayload.put("context", context);

        eventPublisher.publishEvent("order.error", errorPayload, traceId);

        ApiResponse<T> response = ApiResponse.error(message, e.getMessage(), traceId);

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
    }

    private String getStackTrace(Exception e) {
        StringBuilder sb = new StringBuilder();
        for (StackTraceElement element : e.getStackTrace()) {
            sb.append(element.toString()).append("\n");
        }
        return sb.toString();
    }
}
